File: plugins/keyboard/ios/templates/UserLexicon.swift
import Foundation

/// The keyboard's learned vocabulary (§5.11): words the user actually types
/// ("figma", "sats", teammate names) plus a blacklist of autocorrects the user
/// reverted, so the same wrong fix is never auto-applied twice. Persisted as
/// JSON in the App Group container — no Full Access, no network.
///
/// Learning rules (mirrored by `UserLexicon.kt`):
///  - an unknown word committed with a separator twice is learned,
///  - tapping the verbatim strip slot learns immediately,
///  - reverting an autocorrect blacklists that exact typed→corrected pair.
final class UserLexicon {

    private struct Entry: Codable {
        var c: Int // commit count
        var t: TimeInterval // last-used epoch seconds
    }

    private struct Store: Codable {
        var version: Int = 1
        var words: [String: Entry] = [:]
        var blacklist: [String: TimeInterval] = [:]
        /// Word pairs the user actually types, keyed "prev next" — feeds
        /// next-word prediction ahead of the static bigram table.
        var bigrams: [String: Entry] = [:]
    }

    static let maxWords = 5000
    static let maxBlacklist = 500
    static let maxBigrams = 2000
    private static let learnAfterCommits = 2
    private static let flushAfterMutations = 20
    private static let fileName = "keyboard-user-lexicon.json"

    private var store = Store()
    /// Unknown words seen once; promoted into `store.words` on the second
    /// commit. In-memory only — a word must prove itself within one session.
    private var pendingWords: [String: Int] = [:]
    private var mutationsSinceFlush = 0
    private var loaded = false

    private let fileURL: URL? = {
        guard
            let container = FileManager.default.containerURL(
                forSecurityApplicationGroupIdentifier: KeyboardSettings.appGroupID
            )
        else { return nil }
        let dir = container.appendingPathComponent("Library", isDirectory: true)
        try? FileManager.default.createDirectory(
            at: dir, withIntermediateDirectories: true
        )
        return dir.appendingPathComponent(UserLexicon.fileName)
    }()

    // MARK: - Persistence

    /// Called off the main thread at engine init. A missing or unparseable
    /// file starts fresh — learned words are a cache, never precious.
    func load() {
        defer { loaded = true }
        guard
            let url = fileURL,
            let data = try? Data(contentsOf: url),
            let decoded = try? JSONDecoder().decode(Store.self, from: data)
        else { return }
        store = decoded
    }

    /// Writes synchronously; callers flush from `viewWillDisappear` or after
    /// a burst of mutations, never on the keystroke path.
    func flush() {
        guard loaded, mutationsSinceFlush > 0, let url = fileURL else { return }
        mutationsSinceFlush = 0
        guard let data = try? JSONEncoder().encode(store) else { return }
        try? data.write(to: url, options: .atomic)
    }

    private func markMutated() {
        mutationsSinceFlush += 1
        if mutationsSinceFlush >= Self.flushAfterMutations {
            flush()
        }
    }

    // MARK: - Queries

    func contains(_ word: String) -> Bool {
        loaded && store.words[word.lowercased()] != nil
    }

    /// Candidate weight for a learned word, comparable to the dictionary's
    /// log-quantized byte: starts a bit below common words and grows with use.
    func freqQ(_ word: String) -> UInt8? {
        guard loaded, let entry = store.words[word.lowercased()] else { return nil }
        return UInt8(min(255, 96 + 16 * entry.c))
    }

    func isBlacklisted(typed: String, corrected: String) -> Bool {
        loaded && store.blacklist[Self.pairKey(typed, corrected)] != nil
    }

    /// All learned words — the correction engine fuzzy-matches against these
    /// alongside the bundled dictionary (≤5000 words, trivially fast).
    func allWords() -> [String] {
        loaded ? Array(store.words.keys) : []
    }

    /// Learned continuations of `previous`, most-used first. Linear scan over
    /// ≤2000 entries — trivially fast.
    func nextWords(after previous: String, limit: Int = 3) -> [String] {
        guard loaded else { return [] }
        let prefix = previous.lowercased() + " "
        return store.bigrams
            .filter { $0.key.hasPrefix(prefix) }
            .sorted { $0.value.c != $1.value.c ? $0.value.c > $1.value.c : $0.key < $1.key }
            .prefix(limit)
            .map { String($0.key.dropFirst(prefix.count)) }
    }

    // MARK: - Learning

    /// Feed every separator-committed word through here. Unknown words are
    /// learned on their second commit; known (dictionary or learned) words
    /// bump their use count so their suggestions strengthen.
    func observeCommit(word: String, isInDictionary: Bool) {
        guard loaded else { return }
        let key = word.lowercased()
        guard Self.isLearnable(key) else { return }
        if store.words[key] != nil {
            bump(key)
            return
        }
        if isInDictionary { return }
        let seen = (pendingWords[key] ?? 0) + 1
        if seen >= Self.learnAfterCommits {
            pendingWords.removeValue(forKey: key)
            learnNow(word)
        } else {
            pendingWords[key] = seen
        }
    }

    /// Immediate learn — the user explicitly kept the typed word by tapping
    /// the verbatim slot.
    func learnNow(_ word: String) {
        guard loaded else { return }
        let key = word.lowercased()
        guard Self.isLearnable(key) else { return }
        if store.words[key] != nil {
            bump(key)
            return
        }
        evictIfNeeded()
        store.words[key] = Entry(c: Self.learnAfterCommits, t: Self.now())
        markMutated()
    }

    /// Records a committed word pair for next-word prediction. Call with
    /// vetted words only (both known to the dictionary or lexicon) so typos
    /// never become predictions.
    func observeBigram(previous: String, word: String) {
        guard loaded else { return }
        let prev = previous.lowercased()
        let next = word.lowercased()
        // Unlike learned words, a 1-char prev is fine ("i think", "a coffee").
        guard (1...32).contains(prev.count),
              prev.allSatisfy({ ($0.isLetter && $0.isASCII) || $0 == "'" }),
              Self.isLearnable(next) else { return }
        let key = prev + " " + next
        if var entry = store.bigrams[key] {
            entry.c += 1
            entry.t = Self.now()
            store.bigrams[key] = entry
        } else {
            if store.bigrams.count >= Self.maxBigrams {
                let now = Self.now()
                let victim = store.bigrams.min { a, b in
                    Self.retentionScore(a.value, now: now)
                        < Self.retentionScore(b.value, now: now)
                }
                if let victim {
                    store.bigrams.removeValue(forKey: victim.key)
                }
            }
            store.bigrams[key] = Entry(c: 1, t: Self.now())
        }
        markMutated()
    }

    /// The user backspaced an autocorrect: never auto-apply this exact pair
    /// again (it stays available as a tappable suggestion).
    func recordRevert(typed: String, corrected: String) {
        guard loaded else { return }
        if store.blacklist.count >= Self.maxBlacklist {
            // FIFO by timestamp.
            if let oldest = store.blacklist.min(by: { $0.value < $1.value }) {
                store.blacklist.removeValue(forKey: oldest.key)
            }
        }
        store.blacklist[Self.pairKey(typed, corrected)] = Self.now()
        markMutated()
    }

    // MARK: - Internals

    private func bump(_ key: String) {
        guard var entry = store.words[key] else { return }
        entry.c += 1
        entry.t = Self.now()
        store.words[key] = entry
        markMutated()
    }

    /// Recency-weighted eviction: drop the entry with the lowest
    /// `count * exp(-ageDays / 90)` once the cap is reached.
    private func evictIfNeeded() {
        guard store.words.count >= Self.maxWords else { return }
        let now = Self.now()
        let victim = store.words.min { a, b in
            Self.retentionScore(a.value, now: now) <
                Self.retentionScore(b.value, now: now)
        }
        if let victim {
            store.words.removeValue(forKey: victim.key)
        }
    }

    private static func retentionScore(_ entry: Entry, now: TimeInterval) -> Double {
        let ageDays = max(0, now - entry.t) / 86_400
        return Double(entry.c) * exp(-ageDays / 90)
    }

    private static func isLearnable(_ key: String) -> Bool {
        guard key.count >= 2, key.count <= 32 else { return false }
        return key.allSatisfy { ($0.isLetter && $0.isASCII) || $0 == "'" }
    }

    private static func pairKey(_ typed: String, _ corrected: String) -> String {
        typed.lowercased() + "\u{2192}" + corrected.lowercased()
    }

    private static func now() -> TimeInterval {
        Date().timeIntervalSince1970
    }
}

File: plugins/keyboard/android/templates/UserLexicon.kt
package com.a1lab.echos.ime

import android.content.Context
import java.io.File
import kotlin.math.exp
import kotlin.math.min
import org.json.JSONObject

/**
 * The keyboard's learned vocabulary (§5.11): words the user actually types
 * ("figma", "sats", teammate names) plus a blacklist of autocorrects the user
 * reverted, so the same wrong fix is never auto-applied twice. Persisted as
 * JSON in the app's files directory (same channel as keyboard-settings.json).
 *
 * Learning rules (mirrored by `UserLexicon.swift`):
 *  - an unknown word committed with a separator twice is learned,
 *  - tapping the verbatim strip slot learns immediately,
 *  - reverting an autocorrect blacklists that exact typed→corrected pair.
 */
class UserLexicon(baseDir: File) {

    constructor(context: Context) : this(context.applicationContext.filesDir)

    companion object {
        const val MAX_WORDS = 5000
        const val MAX_BLACKLIST = 500
        const val MAX_BIGRAMS = 2000
        private const val LEARN_AFTER_COMMITS = 2
        private const val FLUSH_AFTER_MUTATIONS = 20
        private const val FILENAME = "keyboard-user-lexicon.json"

        private fun isLearnable(key: String): Boolean {
            if (key.length < 2 || key.length > 32) return false
            return key.all { (it in 'a'..'z') || it == '\'' }
        }

        private fun isBigramHalf(key: String): Boolean {
            // Unlike learned words, a 1-char prev is fine ("i think").
            if (key.isEmpty() || key.length > 32) return false
            return key.all { (it in 'a'..'z') || it == '\'' }
        }
    }

    private data class Entry(var count: Int, var lastUsed: Long)

    private val file = File(baseDir, FILENAME)
    private val words = HashMap<String, Entry>()
    private val blacklist = HashMap<String, Long>()

    /** Word pairs the user actually types, keyed "prev next" — feeds
     *  next-word prediction ahead of the static bigram table. */
    private val bigrams = HashMap<String, Entry>()

    /** Unknown words seen once; promoted into [words] on the second commit.
     *  In-memory only — a word must prove itself within one session. */
    private val pendingWords = HashMap<String, Int>()
    private var mutationsSinceFlush = 0
    @Volatile
    private var loaded = false

    // -- Persistence --

    /** Called off the main thread at engine init. A missing or unparseable
     *  file starts fresh — learned words are a cache, never precious. */
    fun load() {
        try {
            if (file.exists()) {
                val json = JSONObject(file.readText(Charsets.UTF_8))
                val wordsJson = json.optJSONObject("words") ?: JSONObject()
                for (key in wordsJson.keys()) {
                    val entry = wordsJson.optJSONObject(key) ?: continue
                    words[key] = Entry(entry.optInt("c", 1), entry.optLong("t", 0L))
                }
                val blacklistJson = json.optJSONObject("blacklist") ?: JSONObject()
                for (key in blacklistJson.keys()) {
                    blacklist[key] = blacklistJson.optLong(key, 0L)
                }
                val bigramsJson = json.optJSONObject("bigrams") ?: JSONObject()
                for (key in bigramsJson.keys()) {
                    val entry = bigramsJson.optJSONObject(key) ?: continue
                    bigrams[key] = Entry(entry.optInt("c", 1), entry.optLong("t", 0L))
                }
            }
        } catch (_: Exception) {
            words.clear()
            blacklist.clear()
            bigrams.clear()
        }
        loaded = true
    }

    /** Writes synchronously; callers flush from `onFinishInputView` or after
     *  a burst of mutations, never on the keystroke path. */
    fun flush() {
        if (!loaded || mutationsSinceFlush == 0) return
        mutationsSinceFlush = 0
        try {
            val wordsJson = JSONObject()
            for ((key, entry) in words) {
                wordsJson.put(
                    key,
                    JSONObject().put("c", entry.count).put("t", entry.lastUsed),
                )
            }
            val blacklistJson = JSONObject()
            for ((key, time) in blacklist) {
                blacklistJson.put(key, time)
            }
            val bigramsJson = JSONObject()
            for ((key, entry) in bigrams) {
                bigramsJson.put(
                    key,
                    JSONObject().put("c", entry.count).put("t", entry.lastUsed),
                )
            }
            val root = JSONObject()
                .put("version", 1)
                .put("words", wordsJson)
                .put("blacklist", blacklistJson)
                .put("bigrams", bigramsJson)
            val tmp = File(file.parentFile, "$FILENAME.tmp")
            tmp.writeText(root.toString(), Charsets.UTF_8)
            tmp.renameTo(file)
        } catch (_: Exception) {
            // Learned words are best-effort; never crash the keyboard.
        }
    }

    private fun markMutated() {
        mutationsSinceFlush += 1
        if (mutationsSinceFlush >= FLUSH_AFTER_MUTATIONS) flush()
    }

    // -- Queries --

    fun contains(word: String): Boolean =
        loaded && words.containsKey(word.lowercase())

    /** Candidate weight for a learned word, comparable to the dictionary's
     *  log-quantized byte: starts a bit below common words, grows with use. */
    fun freqQ(word: String): Int? {
        if (!loaded) return null
        val entry = words[word.lowercase()] ?: return null
        return min(255, 96 + 16 * entry.count)
    }

    fun isBlacklisted(typed: String, corrected: String): Boolean =
        loaded && blacklist.containsKey(pairKey(typed, corrected))

    /** All learned words — the correction engine fuzzy-matches against these
     *  alongside the bundled dictionary (≤5000 words, trivially fast). */
    fun allWords(): List<String> = if (loaded) words.keys.toList() else emptyList()

    /** Learned continuations of [previous], most-used first. Linear scan over
     *  ≤2000 entries — trivially fast. */
    fun nextWords(previous: String, limit: Int = 3): List<String> {
        if (!loaded) return emptyList()
        val prefix = previous.lowercase() + " "
        return bigrams.entries
            .filter { it.key.startsWith(prefix) }
            .sortedWith(
                compareByDescending<Map.Entry<String, Entry>> { it.value.count }
                    .thenBy { it.key },
            )
            .take(limit)
            .map { it.key.removePrefix(prefix) }
    }

    /** Records a committed word pair for next-word prediction. Call with
     *  vetted words only (both known to the dictionary or lexicon) so typos
     *  never become predictions. */
    fun observeBigram(previous: String, word: String) {
        if (!loaded) return
        val prev = previous.lowercase()
        val next = word.lowercase()
        if (!isBigramHalf(prev) || !isLearnable(next)) return
        val key = "$prev $next"
        val existing = bigrams[key]
        if (existing != null) {
            existing.count += 1
            existing.lastUsed = now()
        } else {
            if (bigrams.size >= MAX_BIGRAMS) {
                val nowSeconds = now()
                bigrams.minByOrNull { retentionScore(it.value, nowSeconds) }
                    ?.let { bigrams.remove(it.key) }
            }
            bigrams[key] = Entry(1, now())
        }
        markMutated()
    }

    // -- Learning --

    /** Feed every separator-committed word through here. Unknown words are
     *  learned on their second commit; known (dictionary or learned) words
     *  bump their use count so their suggestions strengthen. */
    fun observeCommit(word: String, isInDictionary: Boolean) {
        if (!loaded) return
        val key = word.lowercase()
        if (!isLearnable(key)) return
        if (words.containsKey(key)) {
            bump(key)
            return
        }
        if (isInDictionary) return
        val seen = (pendingWords[key] ?: 0) + 1
        if (seen >= LEARN_AFTER_COMMITS) {
            pendingWords.remove(key)
            learnNow(word)
        } else {
            pendingWords[key] = seen
        }
    }

    /** Immediate learn — the user explicitly kept the typed word by tapping
     *  the verbatim slot. */
    fun learnNow(word: String) {
        if (!loaded) return
        val key = word.lowercase()
        if (!isLearnable(key)) return
        if (words.containsKey(key)) {
            bump(key)
            return
        }
        evictIfNeeded()
        words[key] = Entry(LEARN_AFTER_COMMITS, now())
        markMutated()
    }

    /** The user backspaced an autocorrect: never auto-apply this exact pair
     *  again (it stays available as a tappable suggestion). */
    fun recordRevert(typed: String, corrected: String) {
        if (!loaded) return
        if (blacklist.size >= MAX_BLACKLIST) {
            blacklist.minByOrNull { it.value }?.let { blacklist.remove(it.key) }
        }
        blacklist[pairKey(typed, corrected)] = now()
        markMutated()
    }

    // -- Internals --

    private fun bump(key: String) {
        val entry = words[key] ?: return
        entry.count += 1
        entry.lastUsed = now()
        markMutated()
    }

    /** Recency-weighted eviction: drop the entry with the lowest
     *  `count * exp(-ageDays / 90)` once the cap is reached. */
    private fun evictIfNeeded() {
        if (words.size < MAX_WORDS) return
        val nowSeconds = now()
        words.minByOrNull { retentionScore(it.value, nowSeconds) }
            ?.let { words.remove(it.key) }
    }

    /** Recency-weighted keep score: usage count decayed by age (90-day
     *  half-life-ish). Lowest score is evicted first. */
    private fun retentionScore(entry: Entry, nowSeconds: Long): Double {
        val ageDays = ((nowSeconds - entry.lastUsed).coerceAtLeast(0L)) / 86_400.0
        return entry.count * exp(-ageDays / 90.0)
    }

    private fun pairKey(typed: String, corrected: String): String =
        typed.lowercase() + "→" + corrected.lowercase()

    private fun now(): Long = System.currentTimeMillis() / 1000
}
